core/src/net/udp_socket.cpp

```diff
@@ -0,0 +1,186 @@
+
+#include <cubos/core/log.hpp>
+#include <cubos/core/net/udp_socket.hpp>
+#include <cubos/core/reflection/external/cstring.hpp>
+#include <cubos/core/reflection/external/primitives.hpp>
+#include <cubos/core/reflection/external/string.hpp>
+
+using cubos::core::net::UdpSocket;
+
+UdpSocket::UdpSocket()
+{
+#ifdef _WIN32
+    WSADATA wsa;
+    CUBOS_ASSERT(WSAStartup(MAKEWORD(2, 2), &wsa) == 0, " WSAStartup failed: {}", WSAGetLastError());
+#endif
+}
+
+UdpSocket::~UdpSocket()
+{
+    unbind();
+}
+
+/// @brief Cross-platform utility to get last error message from stderr.
+/// @return Message.
+/// @todo Move this to a higher module for general purposes?
+static std::string error()
+{
+#ifdef _WIN32
+    static char message[256] = {0};
+    FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, 0, WSAGetLastError(), 0, message, 256, 0);
+    char* nl = strrchr(message, '\n');
+    if (nl)
+        *nl = 0;
+    return {message};
+#else
+    return {strerror(errno)};
+#endif
+}
+
+bool UdpSocket::bind(uint16_t port, const Address& address)
+{
+    mSock = socket(AF_INET, SOCK_DGRAM, 0);
+    if (mSock == INVALID_SOCKET)
+    {
+        CUBOS_ERROR("Failed to allocate UDP socket: {}", error());
+        return false;
+    }
+
+    CUBOS_TRACE("New UDP socket: '{}'", mSock);
+
+    sockaddr_in addr;
+    addr.sin_family = AF_INET;
+    addr.sin_port = htons(port);
+    addr.sin_addr.s_addr = inet_addr(address.toString().c_str());
+
+    if (::bind(mSock, reinterpret_cast<struct sockaddr*>(&addr), sizeof(addr)) == -1)
+    {
+        CUBOS_ERROR("Failed to bind UDP socket: {}", error());
+        unbind();
+        return false;
+    }
+
+    CUBOS_INFO("Bind UDP socket '{}' at address {} with port '{}'", mSock, address.toString(), port);
+    return true;
+}
+
+void UdpSocket::unbind() const
+{
+    if (mSock != INVALID_SOCKET)
+    {
+        CUBOS_WARN("Closing UDP socket: '{}'", mSock);
+#ifdef _WIN32
+        closesocket(mSock);
+        WSACleanup();
+#else
+        close(mSock);
+#endif
+    }
+}
+
+bool UdpSocket::send(const void* data, std::size_t size, const Address& address, uint16_t port) const
+{
+    if (mSock == INVALID_SOCKET)
+    {
+        CUBOS_ERROR("Invalid socket: {}", error());
+        return false;
+    }
+
+    sockaddr_in addr;
+    addr.sin_family = AF_INET;
+    addr.sin_port = htons(port);
+    addr.sin_addr.s_addr = inet_addr(address.toString().c_str());
+
+#ifdef _WIN32
+    auto res = sendto(mSock, static_cast<const char*>(data), static_cast<int>(size), 0,
+                      reinterpret_cast<struct sockaddr*>(&addr), sizeof(addr));
+#else
+    auto res = sendto(mSock, data, size, 0, reinterpret_cast<struct sockaddr*>(&addr), sizeof(addr));
+#endif
+    if (res == INVALID_SOCKET)
+    {
+        CUBOS_ERROR("Failed to send UDP data: {}", error());
+        return false;
+    }
+
+    CUBOS_INFO("Sent UDP message to socket '{}' at address {} and port '{}'", mSock, address.toString(), port);
```

Comment:
This will be very spammy, I would keep it `CUBOS_TRACE` if you find it necessary.